providers/common/provider_ctx.c

```diff
@@ -9,14 +9,31 @@
 
 #include <stdlib.h>
 #include "prov/provider_ctx.h"
+#include "prov/bio.h"
 
 PROV_CTX *PROV_CTX_new(void)
 {
-    return OPENSSL_zalloc(sizeof(PROV_CTX));
+    PROV_CTX *ctx = OPENSSL_zalloc(sizeof(*ctx));
+
+    if (ctx == NULL)
+        return NULL;
+
+# ifndef FIPS_MODULE
+    ctx->corebiometh = bio_prov_init_bio_method();
+    if (ctx->corebiometh == NULL) {
+        OPENSSL_free(ctx);
+        return NULL;
+    }
+# endif
+
+    return ctx;
```

Comment:
This is a case where I believe it would be simpler if each provider init function did this on their own as needed.

``` C
    if ((libctx = [... whatever ...]) == NULL
        || !(tmpprovctx = PROV_CTX_new()) == NULL
        || !(biometh = bio_prov_init_bio_method()) == NULL) {
        BIO_meth_free(biometh);
        PROV_CTX_free(tmpprovctx);
        [... whatever ...](libctx);
        return 0;
    }
    tmpprovctx->handle = provider;
    tmpprovctx->libctx = libctx;
    tmpprovctx->corebiometh = biometh;
    *provctx = tmpprovctx;
```

The internal FIPS provider wouldn't do the biometh part at all.
This is the pattern I set forth in #11803, and it was knowingly done this way to avoid exactly the kind of linking problem you got yourself into...

Reply:
Done